sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala

```diff
@@ -650,4 +650,46 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
       }
     }
   }
+
+  test("Non-partitioned data source table support fallback to HDFS for size estimation") {
+    withTempDir { dir =>
+      Seq(false, true).foreach { fallBackToHDFS =>
+        withSQLConf(SQLConf.ENABLE_FALL_BACK_TO_HDFS_FOR_STATS.key -> s"$fallBackToHDFS") {
+          withTable("spark_25474") {
+            sql(s"CREATE TABLE spark_25474 (c1 BIGINT) USING PARQUET LOCATION '${dir.toURI}'")
+            spark.range(5).write.mode(SaveMode.Overwrite).parquet(dir.getCanonicalPath)
+
+            assert(getCatalogTable("spark_25474").stats.isEmpty)
+            val relation = spark.table("spark_25474").queryExecution.analyzed.children.head
+            // fallBackToHDFS = true: The table stats will be recalculated by DetermineTableStats
+            // fallBackToHDFS = false: The table stats will be recalculated by FileIndex
```

Comment:
Why Spark only do it for non-partitioned tables?

Reply:
- non-partitioned tables uses `PartitioningAwareFileIndex`:

https://github.com/apache/spark/blob/b276788d57b270d455ef6a7c5ed6cf8a74885dde/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala#L103

- partitioned tables uses `CatalogFileIndex`:

https://github.com/apache/spark/blob/5d672b7f3e07cfd7710df319fc6c7d2b9056a068/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CatalogFileIndex.scala#L41

https://github.com/apache/spark/blob/c30b5297bc607ae33cc2fcf624b127942154e559/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L383-L388

Reply:
If so, the behaviour of the non-partitioned case should follow that of the partitioned case? (If `fallBackToHDFS`=false, we should use `defaultSizeInBytes` here, too.

Reply:
no we can't do that, as it will be an regression (image a query picks broadcast join in Spark 2.4 and SMJ in Spark 3.0).

Since partitioned table usually has many many files, I'm OK with the current behavior.